components/theia/packages/gitpod-extension/src/browser/gitpod-user-storage-provider.ts

```diff
@@ -68,31 +74,84 @@ export class GitpodUserStorageProvider implements FileSystemProviderWithFileRead
     }
 
     async readFile(resource: URI): Promise<Uint8Array> {
-        const service = this.getService();
-        const server = service.server;
         // TODO server could resolve binary instead
         const uri = this.toBackwardCompatibleUri(resource);
-        const content = await server.getUserStorageResource({ uri });
+        const content = await this.cachedRead({ uri });
         const value = this.encodingService.encode(content).buffer;
-        this.updateStat(resource, value);
+        this.updateStat(resource, value, false);
         return value;
     }
 
     async writeFile(resource: URI, value: Uint8Array, opts: FileWriteOptions): Promise<void> {
-        const service = this.getService();
-        const server = service.server;
         const content = this.encodingService.decode(BinaryBuffer.wrap(value));
         // TODO server could accept binary instead
         const uri = this.toBackwardCompatibleUri(resource);
-        await server.updateUserStorageResource({ uri, content });
-        this.updateStat(resource, value);
+        await this.cachedWrite({ uri, content });
+        this.updateStat(resource, value, true);
         // TODO server could fire a notification that some resournce was changed instead
         this.onDidChangeFileEmitter.fire([{
             type: FileChangeType.UPDATED,
             resource
         }]);
     }
 
+    // To avoid having badly written extensions executing too much requests and putting pressure on other components we
+    // cache writes and reads here.
+    protected readonly contentCache = new Map<string, string>();
+    protected updatedSinceLastWrite = new Set<string>();
+    protected async cachedWrite(resource: GitpodServer.UpdateUserStorageResourceOptions): Promise<void> {
+        this.contentCache.set(resource.uri, resource.content);
+        this.updatedSinceLastWrite.add(resource.uri);
+        
+        console.count(`user storage write to: ${resource.uri}`);
```

Comment:
i afraid it would be too loud, @AlexTugarev do you really want it?

Reply:
Me too, but I understand Alex wants some insights here... I'd hope the it's not too much per workspace.